pkg/watchdogd/watchdogd.go

```diff
@@ -71,153 +94,266 @@ func MonitorOops() error {
-		// Skip matches where the wildcard is not an integer.
-		pidStr := filepath.Base(filepath.Dir(f))
-		pid, err := strconv.Atoi(pidStr)
-		if err != nil {
-			continue
-		}
-		// Skip matches for the current process.
-		if pid == os.Getpid() {
-			continue
-		}
-		// Skip kernel workers. This is the same mechanism used by ps.
-		cmdline, err := ioutil.ReadFile(filepath.Join(filepath.Dir(f), "cmdline"))
-		if err != nil || len(cmdline) == 0 {
-			continue
+	go func() {
+		log.Println("watchdogd: Start serving.")
+		startServing(l)
+	}()
+
+	log.Println("watchdogd: Start arming watchdog initially.")
+	if r := armWatchdog(); r != OpResultOK {
+		return fmt.Errorf("watchdogd: Initial arm failed")
+	}
+	log.Println("watchdogd: Watchdog armed.")
+
+	if r := startPetting(); r != OpResultOK {
+		return fmt.Errorf("watchdogd: Start petting failed")
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			cleanup()
 		}
-		// /proc files have a gratuitous newline.
-		if string(comm) == DaemonProcess+"\n" {
-			p, err := os.FindProcess(pid)
-			return (*Daemon)(p), err
+	}
+
+}
+
+// doMonitors is a helper function to run the monitors.
+//
+// If there is anything wrong identified, it serves as a signal to stop
+// petting Watchdog.
+func doMonitors(monitors []func() error) error {
+	for _, m := range monitors {
+		if err := m(); err != nil {
+			return err
 		}
 	}
-	return nil, fmt.Errorf("could not find %q", DaemonProcess)
+	// All monitors return normal.
+	return nil
+}
+
+type client struct {
+	Conn *net.UnixConn
+}
+
+func (c *client) Stop() error {
+	return sendAndCheckResult(c.Conn, OpStop)
 }
 
-// Stop stops the daemon. It can be resumed with Continue().
-func (d *Daemon) Stop() error {
-	return (*os.Process)(d).Signal(unix.SIGSTOP)
+func (c *client) Continue() error {
+	return sendAndCheckResult(c.Conn, OpContinue)
 }
 
-// Continue continues the daemon from a previous Stop().
-func (d *Daemon) Continue() error {
-	return (*os.Process)(d).Signal(unix.SIGCONT)
+func (c *client) Disarm() error {
+	return sendAndCheckResult(c.Conn, OpDisarm)
 }
 
-// Disarm sends a signal to the watchdog daemon to disarm.
-func (d *Daemon) Disarm() error {
-	return (*os.Process)(d).Signal(unix.SIGUSR1)
+func (c *client) Arm() error {
+	return sendAndCheckResult(c.Conn, OpArm)
 }
 
-// Arm sends a signal to the watchdog deamon to arm.
-func (d *Daemon) Arm() error {
-	return (*os.Process)(d).Signal(unix.SIGUSR2)
+// sendAndCheckResult sends operation bit and evaluates result.
+func sendAndCheckResult(c *net.UnixConn, op int) error {
+	n, err := c.Write([]byte{byte(op)})
+	if err != nil {
+		return err
+	}
+	if n != 1 {
+		return errors.New("no error; but message not delivered neither")
+	}
+	buf := make([]byte, 1)
+	n, err = c.Read(buf[:])
+	if err != nil {
+		return fmt.Errorf("error reading from server: %v", err)
+	}
+	if n != 1 {
+		return fmt.Errorf("received message from server, but incorrect length")
+	}
+	r := int(buf[0])
+	if r != OpResultOK {
```

Comment:
If the result is a printable character then the code can be %c

Reply:
Also you can see why to make it a printable character: 
var op buf[0]
the zero value of buf[0], in the current code, is a valid opcode and a valid result! We definitely don't want this.

Reply:
ah! bug -- don't messed with the defaults 

Done